Type DataLoader

aawilson opened the pull request:
Add types to the DataLoader. The main drive was to get `batch_load_fn` to allow the type engine to infer the type of `load`, and this accomplishes that. Two unfortunate side-effects:

One is that I don't know how to type it to accept any Sequence type (or, at the bare minimum, `Tuple | List`), because they're generally type-invariant, so even if it's easy to see that `batch_load_fn` will return a sequence type that matches what it gets, pyright still won't allow assignment because DataLoader has to be typed with a Union to also accept the Generic type, and there's no way that I know to separately TypeVar a Generic type and its parameter.

Two is that I don't know how to type instance properties (or, for that matter, Generics) with 3.5-style comment hints, so this will syntax error in 3.5. Happy to convert it over if anyone knows how.

Example of it working here (note the red lines under the `int` types passed to `load` from the hinter in LSP).

![image](https://user-images.githubusercontent.com/57950/192344942-1f00ecd4-89c2-445c-9926-eb0e450bf619.png)

markedwards commented:
Can you remove the reformatting?  Code formatting is a concern for a different PR, and it makes this change difficult to follow and trace.

aawilson committed:
Type DataLoader

Add types to the DataLoader. The main drive was to get batch_load_fn to
allow the type engine to infer the type of load, and this accomplishes
that. Two unfortunate side-effects:

One is that I don't know how to type it to accept any Sequence type (or,
at the bare minimum, Tuple | List), because they're generally
type-invariant, so even if it's easy to see that batch_load_fn will
return a sequence type that matches what it gets, pyright still won't
allow assignment because DataLoader has to be typed with a Union to also
accept the Generic type, and there's no way that I know to separately
TypeVar a Generic type and its parameter.

Two is that I don't know how to type instance properties (or, for that
matter, Generics) with 3.5-style comment hints, so this will syntax
error in 3.5. Happy to convert it over if anyone knows how.

aawilson commented:
@markedwards No problem, I've done so.

aawilson commented:
Diving into usage, I'm running into a big limitation based on my inability to type these as Sequence (thanks mostly to invariance of `list`), which is around how it's very convenient in some cases to write `batch_load_fn` as just a return of an `asyncio.gather(*[iteration over a coroutine accepting keys as arguments])`. This would be a big stumbling block to adding a `Protocol` to `DataLoader` to enforce the shape of `batch_load_fn` there (a thing I would like to do, since it would be very helpful for eliminating a disconnect between `batch_load_fn`'s return type and `load`'s). This mostly just serves to underscore that if anyone knows how to write a `TypeVar` that will either get around `list` invariance or allow binding a typevar on both the container (`list` or `tuple`) _and_ its contained type (`list[SomeReturnType]` etc), that would be a welcome addition to this PR

markedwards commented on .travis.yml:
```diff
@@ -1,7 +1,6 @@
 language: python
 sudo: false
 python:
-- 3.5
```
This Travis config doesn't actually run any more.  I intend to replace this with GitHub actions workflows.  I guess continuing to support python 3.5 is a goal, but supporting typing in 3.5 should not be a goal IMO.

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,82 @@
+)
+from typing_extensions import Protocol, TypeGuard
+
+__version__ = "0.2.1"
```
nitpick -- this PR really shouldn't make any changes that are not required to add typing, so that its really clear what this change does.

aawilson replied:
Back to single quotes it is. I'm more used to working in codebases that are already `black`ed and this one snuck through on my `git add -p` removing the `black` artifacts from before, so I won't take any of these nits personally.

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,82 @@
+    batch: bool = True
+    max_batch_size: Optional[int] = None
+    cache: Optional[bool] = True
+    _queue: List[Loader[KeyT, ReturnT]]
```
Maybe better to do this inline below:
```
self._queue: List[Loader[KeyT, ReturnT]] = []
```

aawilson replied:
I debated this, I couldn't find any guidance on a preference for typing "private" members and it felt better to put all the type declarations here rather than distinguishing, but I don't have a reason to feel strongly enough about that not to stick it inline.

markedwards commented:
I'm working my way through this.  Might not be done today, but I will ping back here when I've finished.

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,82 @@
-def iscoroutinefunctionorpartial(fn):
-    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
 
+class Loader(tuple, Generic[KeyT, ReturnT]):
```
Why is it actually necessary for Loader to be generic?  I think the original `namedtuple('Loader', 'key,future')` can still be used, since this is only used internally.

I think its a good goal for this PR to have zero runtime impact, to ensure no possibility of runtime breakage.

aawilson replied:
Not keeping it Generic loses typing information through _queue, the generic allows it to be inferred throughout the usage in DataLoader. It should mostly only matter inside DataLoader itself, but I haven't prodded it hard enough to be 100% sure about that.

markedwards replied:
Not having it Generic does lose information, but as far as I can tell that type information isn't needed to satisfy the goal of this change, which is to make `DataLoader` fully typed.

Even though the reworking of `Loader` _seems_ benign, I think the goal of this change should be to have absolutely no runtime changes whatsoever in the code, to avoid any possibility of breakage in the wild.  There could be integrations that depend somehow on Loader being a NamedTuple.

If we change the runtime behavior of the library, I think we will have to do a major release, and ideally we can do this as a point release.

aawilson replied:
Yeah, I'm fine with it, it's always easy to tweak later if someone actually runs into it

aawilson replied:
Worth noting that while taking it out I was reminded that 

```
dispatch_queue_batch
failed_dispatch
```

both use the generic in their signature, and these have public function naming convention even though I can't fathom a reason why they'd actually be called publicly.

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,82 @@
+        batch: Optional[bool] = None,
+        max_batch_size: Optional[int] = None,
+        cache: Optional[bool] = None,
+        get_cache_key: Optional[Callable[[KeyT], CacheKeyT]] = None,
```
I think this ought to be `Optional[Callable[[KeyT], Union[CacheKeyT, KeyT]]]`, otherwise a type error occurs on this:
```
self.get_cache_key = get_cache_key or (lambda x: x)
```

aawilson replied:
Yeah that tracks, I'll do a thing

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,82 @@
 
-class DataLoader(object):
+    key: KeyT
+    future: "Future[ReturnT]"
```
This doesn't need to be quoted, does it?  That goes for most of the type hints added here.

aawilson replied:
It does, because `Future` isn't subscriptable. This is still more specific than using `typing.Awaitable`, given how the library works (and it accepts the type parameters just fine)

![image](https://user-images.githubusercontent.com/57950/193991054-bc42493e-9f81-4620-b01c-6f18860201f0.png)

markedwards replied:
Got it, I didn't notice because `Future` is subscriptable as of Python 3.9.  I would remove quoting on type hints in other cases unless it is necessary.  Or do you think quoting is generally preferred for some reason?

aawilson replied:
In general I try to only quote when it's necessary (for example, inside DataLoader methods, the DataLoader self-reference needs to be quoted since at runtime it doesn't exist when the method is defined (when it isn't already DataLoaderT, but in that case DataLoaderT's bound argument gets quoted))

markedwards commented on aiodataloader.py:
```diff
@@ -117,7 +170,7 @@ def load_many(self, keys):
         return gather(*[self.load(key) for key in keys])
 
-    def clear(self, key):
+    def clear(self, key: KeyT) -> "DataLoader":
```
I'm pretty sure a TypeVar is needed here, like:
```
def clear(self: DataLoaderT, key: KeyT) -> DataLoaderT:
```

That TypeVar would be something like:
```
DataLoaderT = TypeVar("DataLoaderT", bound="DataLoader[Any, Any]")
```

aawilson replied:
Good insight, I'll put that in

aawilson replied:
Actually, I don't know what pyright etc will infer when returning self, lemme do an experiment

aawilson replied:
Can confirm, it's DataLoader[Unknown, Unknown], which honestly seems like something pyright/etc should be able to figure out on its own but I'm finding that life in Python types is like that

markedwards commented on aiodataloader.py:
```diff
@@ -192,7 +246,9 @@ def dispatch_queue(loader):
-async def dispatch_queue_batch(loader, queue):
+async def dispatch_queue_batch(
+    loader: DataLoader[KeyT, ReturnT], queue: List[Loader[KeyT, ReturnT]]
+) -> None:
```
nitpick -- this can stay one line

markedwards commented on test_aiodataloader.py:
```diff
@@ -1,36 +1,45 @@
-    async def fn(keys):
+    async def fn(keys: List) -> List:
         load_calls.append(keys)
         return await resolve(keys)
```
I get a `"None" not callable` type error here.  Maybe get rid of the redefinition of resolve above, and make this:
```
return await (resolve or default_resolve)(keys)
```

aawilson replied:
Hmm this should be narrowed by `if resolve is None` and I'm not getting that error, which type inference engine are you using (and what version)?

aawilson replied:
(that said your suggestion is also fine so if it makes it work in more linters then that's probably the better option anyway)

markedwards commented on test_aiodataloader.py:
```diff
@@ -357,8 +368,12 @@ def do_resolve(x):
+        return await deep_loader.load(tuple(keys))
+
+    a_loader, a_load_calls = id_loader(resolve=do_resolve)
+    b_loader, b_load_calls = id_loader(resolve=do_resolve)
```
Why change this?

aawilson replied:
Making it a lambda caused it to fail `resolve`'s type check in `id_loader` because they're not returning `Coroutine`s

markedwards commented:
The current approach does not work for the subclassing pattern (which is the primary usage described in the README):

```python
class DemoDataLoader(DataLoader[str, Optional[str]]):

    async def batch_load_fn(self, keys: list[str]) -> list[Optional[str]]:  # error: Signature of "batch_load_fn" incompatible with supertype "DataLoader"
        pass
```

I haven't worked out how to solve this yet, but I think this usage needs to work, in addition to passing a `batch_load_fn` in the constructor.

markedwards commented:
> One is that I don't know how to type it to accept any Sequence type (or, at the bare minimum, `Tuple | List`), because they're generally type-invariant, so even if it's easy to see that `batch_load_fn` will return a sequence type that matches what it gets, pyright still won't allow assignment because DataLoader has to be typed with a Union to also accept the Generic type, and there's no way that I know to separately TypeVar a Generic type and its parameter.

I'm not sure what you mean here.  Are you saying you want the return type of `batch_load_fn` to be `List[Sequence[...]]` and that doesn't work?  I get no type errors when doing this:

```
async def batch_load_fn(keys: list[str]) -> list[Optional[Sequence[int]]]:
    ...


dl = DataLoader(batch_load_fn)
```

And the revealed type of dl is `aiodataloader.DataLoader[builtins.str, typing.Sequence[builtins.int]]`

markedwards commented:
Oh I see what you mean, you want to change the typing of `BatchLoadFnProto` to have a more permissive return type.

What about ditching the Protocol and typing the `batch_load_fn` argument in the constructor like this:

```
batch_load_fn: Optional[Callable[[List[KeyT]], Coroutine[Any, Any, Sequence[ReturnT]]]] = None,
```

Could even be `Coroutine[Any, Any, Iterable[ReturnT]]` I think.

This also seems to resolve my earlier comment about supporting the subclassing pattern.

markedwards commented:
Note -- this PR should also add a `py.typed` file, to indicate that the library supports type checking.

markedwards commented:
I would suggest we add the following to setup.cfg:
```
[mypy]
files = aiodataloader.py, test_aiodataloader.py
follow_imports = silent
ignore_missing_imports = True
strict = True
```

The `strict = True` will require a bit more work on the typing, but I would argue its worth it.

markedwards commented:
`TypeGuard` should be imported from `typing` if possible:
```python
try:
    from typing import TypeGuard
except ImportError:  # Python < 3.10
    from typing_extensions import TypeGuard
```

markedwards commented:
I also wonder if we need to add `python_requires='>=3.6'` to setup.py, since this change drops support for 3.5.  And edit the classifiers.

aawilson committed:
Tweaks after review

aawilson commented:
> `TypeGuard` should be imported from `typing` if possible:
> 
> ```python
> try:
>     from typing import TypeGuard
> except ImportError:  # Python < 3.10
>     from typing_extensions import TypeGuard
> ```

Is there a way to annotate this so that linters don't give "Unknown import symbol: TypeGuard" on the import within the block? Other than `# type: ignore` that is

EDIT: Ah I guess that's the point of "ignore missing imports" in setup.cfg above

aawilson commented:
> The current approach does not work for the subclassing pattern (which is the primary usage described in the README):
> 
> ```python
> class DemoDataLoader(DataLoader[str, Optional[str]]):
> 
>     async def batch_load_fn(self, keys: list[str]) -> list[Optional[str]]:  # error: Signature of "batch_load_fn" incompatible with supertype "DataLoader"
>         pass
> ```
> 
> I haven't worked out how to solve this yet, but I think this usage needs to work, in addition to passing a `batch_load_fn` in the constructor.

Strange that you're running into the problem, because using my branch at work we are actually using the subclassing pattern and it types fine. I'll poke it more to see what's going on, here.

Just to be clear with my intention (not because I think we disagree), what I believe we want is a situation where the linter can infer the return values of `load` whether `batch_load_fn` is passed to `DataLoader.__init__` or defined as a method on a subclass.

aawilson commented:
> Oh I see what you mean, you want to change the typing of `BatchLoadFnProto` to have a more permissive return type.
> 
> What about ditching the Protocol and typing the `batch_load_fn` argument in the constructor like this:
> 
> ```
> batch_load_fn: Optional[Callable[[List[KeyT]], Coroutine[Any, Any, Sequence[ReturnT]]]] = None,
> ```
> 
> Could even be `Coroutine[Any, Any, Iterable[ReturnT]]` I think.
> 
> This also seems to resolve my earlier comment about supporting the subclassing pattern.

So whenever I typed with Sequence (or Iterable, or Union[list, tuple]), it would fail typing because it would treat Sequence[ReturnT] would not be assignable to a List[ReturnT] type (because List is invariant). Essentially, without a "nested TypeVar" the linter was unable or unwilling to infer that the Sequence[ReturnT] type would be assignable to whatever type the consumer was expecting, because of `list`'s invariance . Maybe this has a different relationship in 3.9+. 

While we're on the subject,  Iterable and Sequence are probably technically a little too loose, since those include str and bytes types that aren't appropriate. I'm not sure consumers of this library would ever really run into that, since the sequence types aren't the inputs to `load`, so if a solution works with them then it's probably good enough.

aawilson commented:
> Note -- this PR should also add a `py.typed` file, to indicate that the library supports type checking.

Looked into this, and there's a [complication](https://peps.python.org/pep-0561/):

> The single-file module should be refactored into a package and indicate that the package supports typing as described above.

So as I understand that'll mean making an aiodataloader subdirectory (probably with a splat import in `__init__.py` to keep usage identical). Should I proceed?

markedwards commented:
@aawilson will catch up on this asap do more reviewing.  Meanwhile, I've merged in a PR that replaces Travis with Github workflows, which means linting and tests run now.

There are some merge conflicts for you to resolve, but shouldn't be difficult.

aawilson committed:
Python > 3.6, and various post-review changes

markedwards commented:
@aawilson I apologize my attention has slowed down on this.  Other priorities have taken precedence, but I promise I will get back to work on this shortly.

markedwards commented on .travis.yml:
```diff
@@ -0,0 +1,22 @@
+    script: flake8
+  - python: '3.7'
+    install: pip install mypy
+    script: mypy aiodataloader.py  --check-untyped-defs  --ignore-missing-imports
```
This file should be removed.  We're not using Travis any longer.

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,66 @@
+except ImportError:  # Python < 3.10
+    from typing_extensions import TypeGuard
+
+from typing_extensions import Protocol
```
Protocol should also be imported directly from typing if possible.  This approach avoids the weird mypy failure:

```python
if sys.version_info >= (3, 8):
    from typing import Protocol
else:
    from typing_extensions import Protocol

if sys.version_info >= (3, 10):
    from typing import TypeGuard
else:
    from typing_extensions import TypeGuard
```

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,66 @@
 
-def iscoroutinefunctionorpartial(fn):
+def iscoroutinefunctionorpartial(fn: Callable) -> TypeGuard[Callable[..., Coroutine]]:
     return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
```
This needs to implement type parameters:
```
def iscoroutinefunctionorpartial(
    fn: Union[Callable[..., ReturnT], partial[ReturnT]],
) -> TypeGuard[Callable[..., Coroutine[Any, Any, ReturnT]]]:
    return iscoroutinefunction(fn.func if isinstance(fn, partial) else fn)
```

Move the TypeVar declarations to the top so you have them everywhere.

aawilson replied:
I remember defining this way at first and then removing the parameterization at some point for some reason, but I don't remember what the reason was, I'll see if it pops up again or I was just mistaken

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,66 @@
+KeyT = TypeVar("KeyT")
+ReturnT = TypeVar("ReturnT")
+CacheKeyT = TypeVar("CacheKeyT")
+DataLoaderT = TypeVar("DataLoaderT", bound="DataLoader")
```
Need type parameters here:
```python
DataLoaderT = TypeVar("DataLoaderT", bound="DataLoader[Any, Any]")
```

aawilson replied:
Is this a mypy complaint? It should be implicitly `Any, Any` but maybe that changed in a python version

markedwards replied:
strict checking does not permit implicit type parameters

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,66 @@
+class BatchLoadFnProto(Protocol[KeyT, ReturnT]):
+    async def __call__(self, keys: List[KeyT]) -> List[ReturnT]:
+        ...
+
```
Need another line here to satisfy flake8

markedwards commented on aiodataloader.py:
```diff
@@ -1,29 +1,66 @@
+        max_batch_size: Optional[int] = None,
+        cache: Optional[bool] = None,
+        get_cache_key: Optional[Callable[[KeyT], Union[CacheKeyT, KeyT]]] = None,
+        cache_map: Optional[Dict[Union[CacheKeyT, KeyT], Any]] = None,
```
I believe this needs to be:

```
cache_map: Optional[Dict[Union[CacheKeyT, KeyT], "Future[ReturnT]"]] = None,
```

markedwards commented on aiodataloader.py:
```diff
@@ -157,18 +194,19 @@ def prime(self, key, value):
 
-def enqueue_post_future_job(loop, loader):
-    async def dispatch():
+def enqueue_post_future_job(loop: AbstractEventLoop, loader: DataLoader) -> None:
```
Needs type parameters: `DataLoader[Any, Any]`

markedwards commented on aiodataloader.py:
```diff
@@ -157,18 +194,19 @@ def prime(self, key, value):
 
 
-def dispatch_queue(loader):
+def dispatch_queue(loader: DataLoader) -> None:
```
Needs type parameters: `DataLoader[Any, Any]`

markedwards commented on aiodataloader.py:
```diff
@@ -192,7 +230,7 @@ def dispatch_queue(loader):
 
 
-async def dispatch_queue_batch(loader, queue):
+async def dispatch_queue_batch(loader: DataLoader, queue: List[Loader]) -> None:
```
Needs type parameters: `DataLoader[Any, Any]`

markedwards commented on aiodataloader.py:
```diff
@@ -243,7 +281,7 @@ async def dispatch_queue_batch(loader, queue):
 
 
-def failed_dispatch(loader, queue, error):
+def failed_dispatch(loader: DataLoader, queue: List[Loader], error: Exception) -> None:
```
Needs type parameters: `DataLoader[Any, Any]`

markedwards commented on setup.cfg:
```diff
@@ -6,3 +6,4 @@ max-line-length = 120
 files = aiodataloader.py, test_aiodataloader.py
 follow_imports = silent
 ignore_missing_imports = True
+strict = True
```
👍

markedwards reviewed (changes requested):
Thanks for your great work here, this is pretty much ready to go.  I have outlined remaining changes, which are minor.

I think this should be release 0.3.0, since it adds functionality in a backwards-compatible way.  I don't think dropping support for python 3.5 warrants a major release.

Regarding the `py.typed`, feel free to make that, or I can do it in a follow-up diff prior to the next release.

In addition to my comments here, `test_aiodataloader.py` needs typing fixes.  I can make another review to help with that if you want.  Do you want to take a crack at it?

aawilson commented:
I happily have a good chunk of time today where I should be able to crank this out. I'd prefer the py.typed change to go into its own branch/pr for cleanliness, if that's okay.

aawilson commented:
I believe the `_tests.py` problems should all be resolved by importing Coroutine and Callable from typing rather than collections.abc, can blame my IDE's auto-importer for that one.

aawilson commented:
Ok, I've bulled through the code quality fixes. I had to do one thing I don't think is controversial (add `# type: ignore` statements to all tests that e.g. alternate exceptions with integers rather than typing them to accept Exceptions since the typing as-is more closely reflects real-world usage), and another thing that I'm surprised I had to do (type the return of id_loader as a single variable Tuple and then destructure it on a new line after it, rather than destructuring it in one line as before)--if there's a way to type-annotate destructured returns without breaking my linter's brain I'd love to hear it. I squashed them all into a single commit [here](https://github.com/syrusakbary/aiodataloader/pull/25/commits/3d3e14f7052169be81d8043ab30cbf3a7a11095d)

markedwards commented:
Thanks @aawilson.  I see checks pass.  I’ll look in detail tomorrow (I’m in London), and hopefully we can wrap this up.

Appreciate your perseverance here :-)

markedwards commented:
I think the `#type: ignore` is a fine solution for a test that intentionally violates the object structure.

Regarding the tuple destructuring with typing, one idea would be to use `typing.cast`, like this:

```
loader_result: Tuple[DataLoader[int, int], List[List[int]]] = id_loader(max_batch_size=2)
identity_loader, load_calls = loader_result
```

becomes

```
identity_loader, load_calls = cast(Tuple[DataLoader[int, int], List[List[int]]], id_loader(max_batch_size=2))
```

I'll hold off until you have a chance to consider this option.  But this PR looks ready to me.

Thanks!!

aawilson commented:
Cast seems like it would potentially mask any typing problems. I'm fine leaving it as it is if you are.

markedwards commented:
> 

Actually it will refuse to cast an incompatible type, and it will also error if the cast isn’t needed.  It should be fine I think.

I don’t strongly object to how it is now though.
